insecure/corruptible/factory.go

```diff
@@ -2,44 +2,111 @@ package corruptible
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"io"
+	"net"
+	"sync"
 
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/rs/zerolog"
 	"google.golang.org/grpc"
+	grpcinsecure "google.golang.org/grpc/credentials/insecure"
 
 	"github.com/onflow/flow-go/insecure"
 	"github.com/onflow/flow-go/model/flow"
 	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
 	"github.com/onflow/flow-go/network"
 )
 
+const networkingProtocolTCP = "tcp"
+
 // ConduitFactory implements a corruptible conduit factory, that creates corruptible conduits and acts as their master.
 // A remote attacker can register itself to this conduit factory.
 // Whenever any corruptible conduit generated by this factory receives an event from its engine, it relays the event to this
 // factory, which in turn is relayed to the register attacker.
 // The attacker can asynchronously dictate the conduit factory to send messages on behalf of the node this factory resides on.
 type ConduitFactory struct {
-	*component.ComponentManager
+	component.Component
+	cm                    *component.ComponentManager
 	logger                zerolog.Logger
 	codec                 network.Codec
 	myId                  flow.Identifier
 	adapter               network.Adapter
 	attackerObserveClient insecure.Attacker_ObserveClient
+	server                *grpc.Server // touch point of attack network to this factory.
+	address               net.Addr
+	ctx                   context.Context
 }
 
-func NewCorruptibleConduitFactory(logger zerolog.Logger, chainId flow.ChainID, myId flow.Identifier, codec network.Codec) *ConduitFactory {
+func NewCorruptibleConduitFactory(
+	logger zerolog.Logger,
+	chainId flow.ChainID,
+	myId flow.Identifier,
+	codec network.Codec,
+	address string) *ConduitFactory {
+
 	if chainId != flow.BftTestnet {
 		panic("illegal chain id for using corruptible conduit factory")
 	}
 
-	return &ConduitFactory{
-		ComponentManager: component.NewComponentManagerBuilder().Build(),
-		myId:             myId,
-		codec:            codec,
-		logger:           logger.With().Str("module", "corruptible-conduit-factory").Logger(),
+	factory := &ConduitFactory{
+		myId:   myId,
+		codec:  codec,
+		logger: logger.With().Str("module", "corruptible-conduit-factory").Logger(),
+	}
+
+	cm := component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			factory.start(ctx, address)
+			factory.ctx = ctx
```

Comment:
why store the context on the factory instead of using the one passed into start?

Reply:
please refer to my answer to the next comment.